### PR TITLE
fix: block.put options

### LINKED ISSUE #815: block.put ignores CID argument
Hello, I'm having an issue where `block.get` is hanging and never returns the block.

Here's the example code:

```js
const ipfsAPI = require('ipfs-api')
const CID = require('cids')
const multihashes = require('multihashes')
const crypto = require('crypto')

function sha256(data) {
  return crypto.createHash('sha256').update(data).digest()
}

const ipfs = ipfsAPI('localhost', '5001', {protocol: 'http'})
const buf = sha256('hello world!')
const mh = multihashes.encode(buf, 'sha2-256')
const cid = new CID(1, 'dag-cbor', mh)

console.log('cid', cid.toBaseEncodedString())

ipfs.repo.gc()
.then(() => {
  return ipfs.block.put(buf, cid)
  .then(block => {
    console.log('put', block.cid.toBaseEncodedString())

    return ipfs.block.get(cid)
    .then(block => {
      // Never gets called!
      console.log('get', block._data.toString())
    })
    .catch(error => console.error(error))
  })
})
```

Output

```bash
cid zdpuAtJJwFer98hHGdt2UQdoF1We7vQuBXSTDKjorNkYovyzQ
put QmPdbXGenToPFzmxJyoVNAAfh5eKvnu2f8KzSx2J8jhwQP
```

It works using the CLI though:

```
$ ipfs repo gc

$ echo -n 'hello world!' | ipfs block put -f cbor
zdpuAtJJwFer98hHGdt2UQdoF1We7vQuBXSTDKjorNkYovyzQ

$ ipfs block get zdpuAtJJwFer98hHGdt2UQdoF1We7vQuBXSTDKjorNkYovyzQ
hello world!
```

Any help is appreciated, thanks.

### PR DESCRIPTION
**Pull request by alanshaw**
Alternative to #793 

Should fix this according to [the spec](https://github.com/ipfs/interface-ipfs-core/blob/master/SPEC/BLOCK.md#blockput).

fixes #815 
closes #793
~~tested by https://github.com/ipfs/interface-ipfs-core/pull/370~~ (released)

### CODE REVIEW

**Comment by alanshaw** [comment]
Please note that the code has a temporary workaround for https://github.com/ipfs/go-cid/issues/75

File: src/block/put.js [hunk 1]
```diff
@@ -3,35 +3,72 @@
...
-    if (typeof cid === 'function') {
-      _callback = cid
-      cid = {}
+  return promisify((block, options, callback) => {
+    if (typeof options === 'function') {
+      callback = options
+      options = {}
     }
 
-    const callback = once(_callback)
```
Comment (diasdavid) [inline src/block/put.js]: I can't remember but I wonder if this was a protection to go-ipfs sometimes sending multiple errors ?!

Comment (alanshaw) [reply src/block/put.js]: Noted, I guess we'll find out soon ;)

File: src/block/put.js [hunk 2]
```diff
@@ -55,7 +55,17 @@ module.exports = (send) => {
 
     sendOneFile(block.data, { qs }, (err, result) => {
       if (err) {
-        return callback(err) // early
+        // Retry with "protobuf" format for go-ipfs
+        // TODO: remove when https://github.com/ipfs/go-cid/issues/75 resolved
+        if (qs.format === 'dag-pb') {
+          qs.format = 'protobuf'
```
Comment (neodon) [inline src/block/put.js]: This is the key change for me that would have allowed me to complete #815.  Still learned a lot from trying to work that story.  Time to find a new issue to dig into. :)

**Pull request merged**